feat(deps): Upgrade phpseclib to v3

susnux opened the pull request:
* For https://github.com/nextcloud/3rdparty/pull/1972

## Summary

Updated to v3 - needed to adjust quiet some places.
AI was just used for https://github.com/nextcloud/server/pull/48183/changes/3983f8a2b80df3f16e549d375d6b49177382833c to help finding a working alternative because the old implementation was using v2 private functions which were not PHP private but just prefixed with `_`.
In v3 they made them really private so we have two options:
- drop our custom streams and use the streams provided by the library
- the solution implemented here using reflection to hack around the private fields...

## Checklist

- Code is [properly formatted](https://docs.nextcloud.com/server/latest/developer_manual/digging_deeper/continuous_integration.html#linting)
- [Sign-off message](https://github.com/src-d/guide/blob/master/developer-community/fix-DCO.md) is added to all commits
- [ ] Tests ([unit](https://docs.nextcloud.com/server/latest/developer_manual/app_development/tutorial.html#unit-tests), [integration](https://docs.nextcloud.com/server/latest/developer_manual/app_development/tutorial.html#integration-tests), api and/or acceptance) are included
- [ ] Screenshots before/after for front-end changes
- [ ] Documentation ([manuals](https://github.com/nextcloud/documentation/) or wiki) has been updated or is not required
- [ ] [Backports requested](https://github.com/nextcloud/backportbot/#usage) where applicable (ex: critical bugfixes)

susnux committed:
feat(deps): Upgrade phpseclib to v3

Signed-off-by: Ferdinand Thiessen <opensource@fthiessen.de>

susnux committed:
feat(deps): Upgrade phpseclib to v3

Signed-off-by: Ferdinand Thiessen <opensource@fthiessen.de>

susnux committed:
fixup: Fix files_external tests

Signed-off-by: Ferdinand Thiessen <opensource@fthiessen.de>

susnux commented:
Problem with this:
`Checker::createSignatureData` does create a different signature... No idea what is happening the parameters are the same.

joshtrichards committed:
fix: adjust hash to v2 library default

Signed-off-by: Josh <josh.t.richards@gmail.com>

joshtrichards committed:
fix: adjust hash to v2 library default

Signed-off-by: Josh <josh.t.richards@gmail.com>

joshtrichards commented on apps/files_external/lib/Lib/Auth/PublicKey/RSAPrivateKey.php:
```diff
@@ -39,12 +39,12 @@ public function __construct(
 			// Add fallback routine for a time where secret was not enforced to be exists
-			$auth->setPassword('');
-			if (!$auth->loadKey($storage->getBackendOption('private_key'))) {
+			$auth->withPassword('');
```
```suggestion
			$auth = $auth->withPassword('');
```

joshtrichards commented on apps/files_external/lib/MountConfig.php:
L220 for v3

	private static function getCipher() {
		$cipher = new AES('cbc');

joshtrichards committed:
fix: set PSS padding / options for RSA keys for v2->v3 consistency 

Updated RSA key loading to use PSS padding and options.

Signed-off-by: Josh <josh.t.richards@gmail.com>

joshtrichards committed:
fix: set PSS padding / options for RSA keys for v2->v3 consistency 

Updated RSA key loading to use PSS padding and options.

Signed-off-by: Josh <josh.t.richards@gmail.com>

terrafrost commented on apps/files_external/lib/Lib/Auth/PublicKey/RSA.php:
```diff
@@ -41,33 +41,33 @@ public function __construct(
-		$auth->setPassword($this->config->getSystemValue('secret', ''));
-		if (!$auth->loadKey($storage->getBackendOption('private_key'))) {
+		try {
+			$auth = RSACrypt::loadPrivateKey(
```
Might be nice to support EC / DSA keys as well. One can do that thusly:

```php
$auth = \phpseclib3\Crypt\PublicKeyLoader::load(
    $storage->getBackendOption('private_key'),
    $this->config->getSystemValue('secret', '');
);
if (!$auth instanceof \phpseclib3\Crypt\Common\PrivateKey) {
    throw new \Exception('A private key needs to be loaded - not a ' . $auth::class . ' key');
}
```
I suppose it might also be worthwhile to realize that if no key can be detected (eg. you try to load a PHP file as a key) then a `\phpseclib3\Exception\NoKeyLoadedException` will be thrown.

That said, if more than just RSA keys are going to be supported then renaming the class would prob be appropriate as well.

susnux replied:
Please open a feature request and if you like also implement that then as a contribution.
We would be happy to review such contribution :)

But for this pull request new features are out-of-scope.

susnux committed:
fix: salt size needs to be 16, 24 or 32 bytes not random 30 bytes

Signed-off-by: Ferdinand Thiessen <opensource@fthiessen.de>

susnux committed:
fix(files_external): adjust custom sftp stream for phpseclib v3

The implementation was using private fields of v2, which are now
properly marked as private.
So either drop the custom streams or use reflection.

Assisted-by: ClaudeCode:claude-opus-4-8
Signed-off-by: Ferdinand Thiessen <opensource@fthiessen.de>

susnux commented on apps/files_external/lib/Lib/Storage/SFTPReflection.php:
Any options on this solution?
Should we keep the "performance" streams with reflection or drop them completely and just use the library streams?

miaulalala reviewed (changes requested):
Did some benchmarking, scripts and instructions in the linked gist below. 

Ran `SFTPReadStream`/`SFTPWriteStream` against plain `SFTP::get()`/`put()` to check whether the pipelining they're meant to provide actually pays off.

### Setup: 
- standalone PHP script, real phpseclib3 v3.0.55 via Composer, the three files from this branch dropped in unmodified, against a local `atmoz/sftp` container. Tested both on localhost and with ~30ms latency injected via `tc netem` inside the container, to approximate a real remote host.

### Results (medians over 5 runs):

localhost, no injected latency, 20 MiB:
- write: public `put()` 247ms vs. reflection stream 9,659ms (~39x slower)
- read: public `get()` 238ms vs. reflection stream 218ms (roughly even)

~30ms injected latency, 1 MiB:
- write: public `put()` 844ms vs. reflection stream 1,641ms (~1.9x slower)
- read: public `get()` 883ms vs. reflection stream 1,173ms (~1.3x slower)

The reflection streams were slower than the public API in pretty much every condition I tried, not faster. I checked this isn't a benchmark artifact (MD5 of the round-tripped file matches the input, so no silent retries).

`SFTPWriteStream::stream_flush()` sends one 64 KiB packet and immediately blocks on `read_put_responses(1)` per chunk. Fully synchronous, not pipelined. 

phpseclib3's own public `put()` already queues up to `NET_SFTP_UPLOAD_QUEUE_SIZE` (1024) packets before batch-reading responses, i.e. it already does the windowed upload the custom stream's docblock claims to add, and does it better. 

`SFTPReadStream` does have one-chunk read-ahead, which is why it's closer, but it still trails `get()` once latency is added.

Script + full setup/repro instructions: https://gist.github.com/miaulalala/ef2ed83521e515984dc26ed8bd60a154

AI was used to create the scripts and benchmark data, so please reconfirm with your own local setup. If those numbers are confirmed for someone else too, I'd vote for dropping the custom streams and using the library's public API.

ETA: added some latency runs in the gist, too, if you're interested.

CarlSchwan reviewed (approved):
I feel like having up to date security relevant dependency is more important than performance
